oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/authentication/DefaultAuthenticationFailureHandler.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2020-2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.server.authorization.web.authentication;
+
+import java.io.IOException;
+
+import jakarta.servlet.ServletException;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.converter.HttpMessageConverter;
+import org.springframework.http.server.ServletServerHttpResponse;
+import org.springframework.security.core.AuthenticationException;
+import org.springframework.security.oauth2.core.OAuth2AuthenticationException;
+import org.springframework.security.oauth2.core.OAuth2Error;
+import org.springframework.security.web.authentication.AuthenticationFailureHandler;
+
+/**
+ * 	A default implementation of an {@link AuthenticationFailureHandler} used for handling an {@link OAuth2AuthenticationException}
+ * 	and returning the {@link OAuth2Error Error Response}.
+ *
+ * @author Dmitriy Dubson
+ * @since 1.2.0
+ * @see AuthenticationFailureHandler
+ */
+public class DefaultAuthenticationFailureHandler implements AuthenticationFailureHandler {
+	private final HttpMessageConverter<OAuth2Error> errorHttpResponseConverter;
+
+	public DefaultAuthenticationFailureHandler(HttpMessageConverter<OAuth2Error> errorHttpResponseConverter) {
+		this.errorHttpResponseConverter = errorHttpResponseConverter;
+	}
+
+	@Override
+	public void onAuthenticationFailure(HttpServletRequest request, HttpServletResponse response, AuthenticationException exception) throws IOException, ServletException {
+		OAuth2Error error = ((OAuth2AuthenticationException) exception).getError();
```

Comment:
We should assert that `exception instanceof OAuth2AuthenticationException` as it's possible this class may be misconfigured resulting in `ClastCastException`. If it's not `OAuth2AuthenticationException`, then log a warning and just return `BAD_REQUEST`.

We should also add a test for this scenario.

Lastly, let's rename parameter `exception` to `authenticationException`

Reply:
Yes, use `logger.warn()` similar to how we log debug:

https://github.com/spring-projects/spring-authorization-server/blob/1ed0b3ada38027af8fc047fdf0ad073a4d0a9c04/oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/ClientSecretAuthenticationProvider.java#L118